packages/dds/tree/docs/undo/README.md

```diff
@@ -169,6 +170,139 @@ Partial undo would allow such a user to only undo the changes within the region
+	/* The commit to undo */
+	readonly commit: GraphCommit<TChange>;
+	/* The repair data associated with the commit */
+	readonly repairData: ReadonlyRepairDataStore;
+	/* The next undoable commit. */
+	readonly parent?: UndoableCommit<TChange>;
+}
+```
+
+That tree is a sparse copy of the commit tree maintained by `EditManager` and `Checkout`s for branch management.
+
+The structure forms a tree as opposed to a linked-list because different local branches can share the same ancestor commits.
+Each branch however only ever sees a single spine of this tree, which therefore looks like a linked-list to said branch.
+The rest of the document uses the term "list" when describing operations performed at the scope of a single branch.
+
+The tree is sparse because it does **_not_** contain the following kinds of edits:
+
+-   Edits authored by other clients (this is only a concern for the part of the tree that represents the trunk).
+-   Undo edits.
+
+Note that some of these edits in the tree may be part of the trunk while others may be on a branch.
+Each branch need only maintain a "head" pointer to the child-most commit on the branch.
+
+### The Redo Commit Tree
+
+The tree of redoable commits is maintained across branches in a similar fashion to the undoable commits tree.
+Redoable commits are effectively undoable commits and can therefore use the same `UndoableCommit` structure described above.
+
+### Reacting to Local Edits
+
+The redo and undo commit lists for a branch are updated as follows in the face of new local edits:
+
+-   If the edit is neither an undo nor a redo:
+    -   A new undoable commit node is pushed onto the undoable commit list.
+        The parent field of the new commit node should point to the previous head undoable commit.
+    -   The head pointer for the redoable commits is cleared for that branch.
+        This effectively drops all the redoable commits for the branch.
+-   If the edit is an undo:
+    -   The head undoable commit is popped.
+        The parent of the undoable commit becomes the new head undoable commit.
+    -   The concrete undo commit is pushed onto the tip of the redoable commit list.
+        The parent field of the new redoable commit node should point to the previous head redoable commit.
+-   If the edit is a redo:
+    -   The head pointer for the redoable commits is popped.
+        The parent of the redoable commit becomes the new head redoable commit.
+    -   The concrete redo commit is pushed onto the undoable commit list.
+        The parent field of the new commit node should point to the previous head undoable commit.
+
+### Forking
+
+When a branch is forked, the new branch can simply obtain the head undoable and redoable commit pointers from the parent branch.
+This helps keep forking cheap.
+
+### Pulling
+
+When a branch is pulled, it must re-attach whatever undoable and redoable commits it has to the head of the undo list maintained by the parent branch.
```

Comment:
When does a branch get pulled?

Reply:
A branch is forked when the app code calls `SharedTreeBranch.pull()` or causes it to be called through other means (e.g., calling `SharedTreeBranch.merge()`).

Note: my answer is assuming [PR14442](https://github.com/microsoft/FluidFramework/pull/14442) goes in, but the gist of the answer is the same:  the pulling itself is not for the purpose of undo, it is imposed from the outside.